gocd/gocd_util.go

```diff
@@ -0,0 +1,280 @@
+package gocd
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+// Server a representstion of a GoCD server
+type Server struct {
+	Host     string
+	Port     string
+	User     string
+	Password string
+	Timeout  time.Duration
+}
+
+// NewServerConfig Create a Server object from a config
+func NewServerConfig(host string, port string, user string, password string, timeoutStr string) *Server {
+
+	// timeout casting to seconds
+	timeout := time.Duration(120 * time.Second)
+	i, err := strconv.Atoi(timeoutStr)
+	if err == nil {
+		timeout = time.Duration(i) * time.Second
+	} else {
+		log.Warn("Failed to convert timeout to seconds: ", err)
+	}
+
+	return &Server{
+		Host:     host,
+		Port:     port,
+		User:     user,
+		Password: password,
+		Timeout:  timeout,
+	}
+}
+
+// URL returns the host of the GoCD server
+func (server Server) URL() string {
+	return fmt.Sprintf("%s:%s", server.Host, server.Port)
+}
+
+// client returns a http client with longer timeout and skip verify
+func client(timeout time.Duration) *http.Client {
+	transCfg := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	}
+	return &http.Client{
+		Timeout:   timeout,
+		Transport: transCfg,
+	}
+}
+
+func printPrettyJSON(body []byte, objectname string) (prettyJSON bytes.Buffer, err error) {
+	err = json.Indent(&prettyJSON, body, "", "\t")
+	if err != nil {
+		log.Warn("Failed to prettify JSON: ", err)
+	}
+	log.Debug(objectname+" JSON:", string(prettyJSON.Bytes()))
+	return
+}
+
+// readPipelineJSONFromFile reads a GoCD structure from a json file
+func readPipelineJSONFromFile(path string) (pipeline Pipeline, err error) {
+	data, err := ioutil.ReadFile(path)
+	if err == nil {
+		err = json.Unmarshal(data, &pipeline)
+	}
+	return
+}
+
+// Partially generated by curl-to-Go: https://mholt.github.io/curl-to-go
+func (server Server) pipelineConfigPUT(pipeline Pipeline, etag string) (pipelineResult Pipeline, err error) {
+
+	pipelineName := pipeline.Name
+
+	payloadBytes, err := json.Marshal(pipeline)
+	if err != nil {
+		return
```

Comment:
Should this error be logged here, or is that handled by the caller?

Reply:
this occurrence, as well as all the "curl-to-Go" generated functions, pass the errors up to the caller.